### PR TITLE
#1920 try to fix race condition on proposal height for published txs

### PR DESCRIPTION
**Pull request by srmo**
Potential fix for https://github.com/tendermint/tendermint/issues/1920

My observation is, that there's a race condition between the actual mempool height used in the propose step and the height used to publish new transactions via `notifyTxsAvailable`

It's possible that `notifyTxsAvailable` publishes a tx with a height, that is already finalized (via votes/short before commiting/`Update()` - thought that's where my understanding ends).

If this happens, the receiveRoutine will drain the tx from the channel and and then call `handleTxsAvailable` which in turn triggers propose. Now the height is just wrong and propose will throw a `enterPropose(229/0): Invalid args. Current step: 230/0/RoundStepNewHeight module=consensus height=229 round=0` and the tx is now stuck in the mempool.
Interestingly, additional broadcasts also do not get published.
Probably because the channel is now full and we hit the `default` branch as the `case` fails to publish.

Calling `handleTx` with the next height (current mempool height + 1) fixes the problem.
I've exposed mempool.height via a new method `Height()` on mempool and added a defensive lock around it - I assume that one has think about remaining race conditions surrounding that.

The root cause analysis is hard for me.
I'm happy to discuss this here / on the issue / ...

- related to create_empty_blocks=false
- published height for accepted tx can be wrong (too low)
- use the actual mempool height + 1 for the proposal
- expose Height() on mempool

* [x] Updated all relevant documentation in docs
* [x] Updated all code comments where relevant
* [ ] Wrote tests
* [x] Updated CHANGELOG.md

### CODE REVIEW

**Comment by srmo** [comment]
and no idea how to increase the MempoolMock coverage

File: consensus/state.go [hunk 1]
```diff
@@ -572,7 +572,9 @@ func (cs *ConsensusState) receiveRoutine(maxSteps int) {
 
 		select {
 		case height := <-cs.mempool.TxsAvailable():
-			cs.handleTxsAvailable(height)
+			actualHeight := cs.mempool.Height() + 1
```
Comment (melekes) [inline consensus/state.go]: can't we use `cs.Height` here?

Comment (melekes) [reply consensus/state.go]: I meant `cs.Height + 1`, sorry

**Comment by srmo** [comment]
Using `cs.Height + 1` didn't work at all. I'm a bit stumped by that but using just `cs.Height` seems to work, at least in a short one-node run. I'll keep the node running over night.

File: consensus/state.go [hunk 2]
```diff
@@ -571,8 +571,11 @@ func (cs *ConsensusState) receiveRoutine(maxSteps int) {
 		var mi msgInfo
 
 		select {
-		case height := <-cs.mempool.TxsAvailable():
-			cs.handleTxsAvailable(height)
+		case txAvailable := <-cs.mempool.TxsAvailable():
+			// TODO why don't we have a "default" branch for this drain? Similar to the publish, wouldn't this lock otherwise?
```
Comment (melekes) [inline consensus/state.go]: this won't lock because we have other cases below

File: consensus/state.go [hunk 3]
```diff
@@ -683,11 +686,13 @@ func (cs *ConsensusState) handleTimeout(ti timeoutInfo, rs cstypes.RoundState) {
 
 }
 
-func (cs *ConsensusState) handleTxsAvailable(height int64) {
+func (cs *ConsensusState) handleTxsAvailable() {
 	cs.mtx.Lock()
 	defer cs.mtx.Unlock()
 	// we only need to do this for round 0
-	cs.enterPropose(height, 0)
+	heightToPropose := cs.Height
```
Comment (melekes) [inline consensus/state.go]: no need to create a separate variable

Comment (srmo) [reply consensus/state.go]: Idea here was being paranoid, as there's still time passing between the call to cs.Height in the logging statement and the actual passing. (and I'm coming from Java where we usually don't access fields but methods, so multiple calls should be consolidated into one variable.
Will remove this.

File: consensus/state.go (hunk 2 above)
Comment (srmo) [reply consensus/state.go]: Ah, OK, understood.

**Comment by srmo** [comment]
I don't know if I need to write an explicit test for this. Any hints or can I tick the checkbox?

**Pull request merged**